src/accounts_refactor.rs

```diff
@@ -0,0 +1,339 @@
+    }
+
+    /// Get the optional ledger scan.
+    pub fn scan(&self) -> Option<BackgroundKeyScan<L>> {
+        self.scan.clone()
+    }
+
+    /// Get the list of asset codes.
+    pub fn assets(&self) -> Vec<AssetCode> {
+        self.assets.clone()
+    }
+
+    /// Get the list of records.
+    pub fn records(&self) -> Vec<RecordInfo> {
+        self.records.clone()
+    }
+
+    /// Get the table of balances.
+    pub fn balances(&self) -> HashMap<AssetCode, U256> {
+        self.balances.clone()
+    }
+
+    /// Get the created time.
+    pub fn created_time(&self) -> DateTime<Local> {
+        self.created_time
+    }
+
+    /// Get the modified time.
+    pub fn modified_time(&self) -> DateTime<Local> {
+        self.modified_time
+    }
+}
+
+pub type AccountsStore<L, Key, PubKey> = KeyValueStore<PubKey, Account<L, Key>>;
+
+/// An editor to create or update the account or accounts store.
+pub struct AccountEditor<'a, L: Ledger, Key: KeyPair + DeserializeOwned + Serialize> {
+    account: Account<L, Key>,
+    store: &'a mut AccountsStore<L, Key, Key::PubKey>,
+}
+
+impl<'a, L: Ledger, Key: KeyPair + DeserializeOwned + Serialize> AccountEditor<'a, L, Key> {
+    /// Create an account editor.
+    fn new(
+        store: &'a mut AccountsStore<L, Key, Key::PubKey>,
+        account: Account<L, Key>,
+    ) -> AccountEditor<'a, L, Key> {
+        Self { account, store }
+    }
+
+    /// Set the account as used.
+    pub fn set_used(mut self) -> Self {
+        self.account.used = true;
+        self.account.modified_time = Local::now();
+        self
+    }
+
+    /// Set the account as unused.
+    pub fn set_unused(mut self) -> Self {
+        self.account.used = false;
+        self.account.modified_time = Local::now();
+        self
+    }
+
+    /// Set the optional ledger scan.
+    pub fn set_scan(mut self, scan: Option<BackgroundKeyScan<L>>) -> Self {
+        self.account.scan = scan;
+        self.account.modified_time = Local::now();
+        self
+    }
+
+    /// Set the ledger scan.
+    pub fn with_scan(mut self, scan: BackgroundKeyScan<L>) -> Self {
+        self.account.scan = Some(scan);
+        self.account.modified_time = Local::now();
+        self
+    }
+
+    /// Clear the leger scan.
+    pub fn clear_scan(mut self) -> Self {
+        self.account.scan = None;
+        self.account.modified_time = Local::now();
+        self
+    }
+
+    /// Update the ledger scan.
+    pub(crate) async fn update_scan(
+        mut self,
+        event: LedgerEvent<L>,
+        source: EventSource,
+        records_commitment: MerkleCommitment,
+    ) -> Result<AccountEditor<'a, L, Key>, KeystoreError<L>> {
+        let mut scan = match self.account.scan.take() {
+            Some(scan) => scan,
+            None => return Err(KeystoreError::ScanNotFound),
+        };
+        scan.handle_event(event, source);
+        // Check if the scan is complete.
+        if let Err(scan) = scan.finalize(records_commitment) {
```

Comment:
This was confusing at first since it's not obvious that Err() isn't really an error condition but just means the scan isn't complete yet.  Also I think we need to return the `ScanOutputs` if the scan is finished.  slightly mess with chaining but we should probably return the output of the scan as well as the editor.

Reply:
Yeah I think it's fine if this method isn't chainable. It's really only going to be called in one place. In fact it might make sense to just move this logic outside the model (e.g. to the controller layer)

Reply:
The comment "Check if the scan is complete" seems to be a bit misleading. `scan.finalize` actually returns an error if the given `records_commitment` doesn't match the scan's Merkle tree, so this line of code is checking if the scan is _successful_, rather than if it's _complete_.

Reply:
The scan is defined to be complete when its Merkle commitment matches the latest one, that's how you know you've caught up to the wallet's main thread. I think there was some discussion of changing this to check block height instead, and then assert that the Merkle commitments are equal, which might be less confusing and more likely to catch errors like the Merkle trees being out of sync, rather than just looping forever.

Also I agree that returning `Result` makes the calling code (here) confusing, since `Err` doesn't necessarily indicate an error case. It's really supposed to be `Either` but there's no `Either` in the standard prelude. We could define an enum like
```rust
ScanStatus {
  Finished { records, history },
  InProgress(BackgroundKeyScan)
}
```